collections/collections.go

```diff
@@ -0,0 +1,98 @@
+package collections
+
+import (
+	"bytes"
+	"fmt"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// StorageProvider represents sdk.Context
+// it is used to avoid to reduce dependencies.
+type StorageProvider interface {
+	// KVStore returns a KVStore given its StoreKey.
+	KVStore(key storetypes.StoreKey) storetypes.KVStore
+}
+
+// ErrNotFound defines an error returned when
+// a key was not found.
+type ErrNotFound struct {
+	// HumanizedKey represents the human-readable key.
+	HumanizedKey string
+	// RawKey represents the raw key in byte format.
+	RawKey []byte
+	// ValueType represents the type which was not found.
+	ValueType string
+}
+
+// Error implements the error interface.
+func (e ErrNotFound) Error() string {
+	return fmt.Sprintf("collections: key not found for %s '%s'", e.ValueType, e.HumanizedKey)
+}
+
+// Is implements error.Is interface
+func (e ErrNotFound) Is(err error) bool {
+	self, ok := err.(ErrNotFound)
+	if !ok {
+		return false
+	}
+	return bytes.Equal(self.RawKey, e.RawKey) &&
+		self.HumanizedKey == e.HumanizedKey &&
+		self.ValueType == e.ValueType
+}
+
+// Namespace provide a storage byte prefix namespace
+// which must be unique across storage objects belonging
+// to a module.
+// Besides this, once a namespace is defined: other module's
+// namespaces cannot start with the same returned prefix to
+// avoid namespace collisions, example:
+// Namespace1=> "bank"
+// Namespace2=> "bank2" <- namespace collision: Namespace1 starts with Namespace2
+type Namespace struct {
+	prefix []byte // TODO(testinginprod): maybe add a humanized prefix?
+}
+
+// Prefix returns the raw prefix bytes.
+func (n Namespace) Prefix() []byte { return n.prefix }
+
+// NewNamespace returns a Namespacer given the provided namespace identifier.
+func NewNamespace[T interface{ uint8 | string | []byte }](identifier T) Namespace {
```

Comment:
this is to support @ValarDragon usecase, we accept a union of types and convert those into raw bytes below.

Reply:
NOTE: golang won't infer uint8 if we do `NewNamespace(1)` so developer will be forced to do `NewNamespace(uint8(1))`, we could use `int` in the union (which is properly inferred) and then panic if the number is out of bounds.

Reply:
swapped uint8 with int and a panicking bound check, this is to avoid the need to write uint8(number).

Reply:
Why not uint64 and varint encode?

Reply:
Although I do think having more than 256 collections in a module would be crazy

Reply:
yes this was my feeling when i initially implemented prefixes/namespaces as uint8, 256 collections in a module seemed like too much